mh/basic.py

```diff
@@ -0,0 +1,374 @@
+
+                                if handler.is_finished():
+                                    self.remove_handler(handler)
+                        finally:
+                            write_watch.restart()
+        finally:
+            self.__is_shut_down.set()
+
+    def _worker_target(self, work_queue):
+        # type: (queue.Queue) -> None
+
+        while not self.__shutdown_request:
+            try:
+                handler, packet, event = work_queue.get(block=True)
+            except queue.Empty:
+                continue
+
+            if self.__shutdown_request:
+                break
+
+            if handler.is_finished():
+                continue
+
+            assert event == selectors.EVENT_READ
+
+            try:
+                handler.on_packet(packet)
+            except Exception as e:
+                handler.on_exception(e)
+
+            if handler.is_finished():
+                self.remove_handler(handler)
+
+    def accept_new_connection(self):
+        # type: () -> None
+
+        try:
+            client_socket, client_address = self.socket.accept()
+        except Exception as e:
+            print('Error accepting connection (1). {}'.format(e))
+            return
+
+        try:
+            handler = self.RequestHandlerClass(client_socket, client_address,
+                                               self)
+        except Exception:
+            print('Error accepting connection (2). {}'.format(e))
+            return
+
+        handler.__worker_thread = \
+            self._random.randint(0, len(self.worker_queues)-1)
+
+        self.selector.register(handler, selectors.EVENT_READ)
+        self.handlers.append(handler)
+
+    def _queue_work(self, handler, work_data, event):
+        # type: (BasicPatHandler, any, int) -> None
+        if handler.is_finished():
+            return
+
+        thread_queue = self.worker_queues[handler.__worker_thread]
+        thread_queue.put((handler, work_data, event), block=True)
+
+    def remove_handler(self, handler):
+        # type: (BasicPatHandler) -> None
+        try:
+            self.handlers.remove(handler)
+        except Exception:
+            pass
+
+        try:
+            self.selector.unregister(handler)
+        except Exception:
+            pass
+
+        try:
+            handler.finish()
+        except Exception:
+            pass
+
+        try:
+            handler.socket.close()
+        except Exception:
+            pass
+
+    def close(self):
+        """Called to clean-up the server.
+
+        May be overridden.
+
+        """
+        self.__shutdown_request = True
+
+        if sys.platform != "win32":
+            # Non Windows selector don't support waking up
+            # the select when a socket is closed, to bypass
+            # that we can just create a new socket that connect
+            # to the server, that way it would wake up the select
+            # and then see that it's closed and be done with it
```

Comment:
Do you have any reference/link to this? That sounds like a really strange bug.

Reply:
This is actually not needed anymore, since we are waking up at least once every time the server thread's `select` timeout (1 second). So, at worst we would be waiting here 1 second before the server thread wake up and set `self.__is_shut_down`